src/signTxPoolRegistration.c

```diff
@@ -101,235 +125,715 @@ static inline void advanceState()
-
-			ASSERT_TYPE(p->cost, uint64_t);
-			p->cost = u8be_read(wireHeader->cost);
-			TRACE_ADA_AMOUNT("cost ", p->cost);
-			VALIDATE(p->cost < LOVELACE_MAX_SUPPLY, ERR_INVALID_DATA);
-
-			ASSERT_TYPE(p->marginNumerator, uint64_t);
-			p->marginNumerator = u8be_read(wireHeader->marginNumerator);
-			TRACE_BUFFER((uint8_t *) &p->marginNumerator, 8);
-			VALIDATE(p->marginNumerator <= MARGIN_DENOMINATOR_MAX, ERR_INVALID_DATA);
-
-			ASSERT_TYPE(p->marginDenominator, uint64_t);
-			p->marginDenominator = u8be_read(wireHeader->marginDenominator);
-			TRACE_BUFFER((uint8_t *) &p->marginDenominator, 8);
-			VALIDATE(p->marginDenominator != 0, ERR_INVALID_DATA);
-			VALIDATE(p->marginDenominator <= MARGIN_DENOMINATOR_MAX, ERR_INVALID_DATA);
-			VALIDATE(p->marginNumerator <= p->marginDenominator, ERR_INVALID_DATA);
-
-			TRACE_BUFFER(wireHeader->rewardAccount, SIZEOF(wireHeader->rewardAccount));
-			STATIC_ASSERT(SIZEOF(wireHeader->rewardAccount) == SIZEOF(p->rewardAccount), "wrong reward account size");
-			os_memmove(p->rewardAccount, wireHeader->rewardAccount, SIZEOF(p->rewardAccount));
-			const uint8_t header = getAddressHeader(p->rewardAccount, SIZEOF(p->rewardAccount));
-			VALIDATE(getAddressType(header) == REWARD, ERR_INVALID_DATA);
-			VALIDATE(getNetworkId(header) == commonTxData->networkId, ERR_INVALID_DATA);
-		}
+__noinline_due_to_stack__
+static void _calculateRewardAccount(
+        const pool_reward_account_t* rewardAccount,
+        uint8_t* rewardAccountBuffer
+)
+{
+	switch (rewardAccount->keyReferenceType) {
 
-		ASSERT_TYPE(subctx->numOwners, uint16_t);
-		ASSERT_TYPE(subctx->numRelays, uint16_t);
-		subctx->numOwners = (uint16_t) u4be_read(wireHeader->numOwners);
-		subctx->numRelays = (uint16_t) u4be_read(wireHeader->numRelays);
+	case KEY_REFERENCE_HASH: {
+		STATIC_ASSERT(SIZEOF(rewardAccount->buffer) == REWARD_ACCOUNT_SIZE, "wrong reward account size");
+		os_memmove(rewardAccountBuffer, rewardAccount->buffer, REWARD_ACCOUNT_SIZE);
+		break;
+	}
+	case KEY_REFERENCE_PATH: {
+		constructRewardAddressFromKeyPath(
+		        &rewardAccount->path, commonTxData->networkId,
+		        rewardAccountBuffer, REWARD_ACCOUNT_SIZE
+		);
+		break;
+	}
+	default:
+		ASSERT(false);
+	}
+}
 
-		TRACE(
-		        "num owners, relays: %d %d",
-		        subctx->numOwners, subctx->numRelays
+enum {
+	HANDLE_POOL_REWARD_ACCOUNT_STEP_DISPLAY_PATH = 6500,
+	HANDLE_POOL_REWARD_ACCOUNT_STEP_DISPLAY_HASH,
+	HANDLE_POOL_REWARD_ACCOUNT_STEP_RESPOND,
+	HANDLE_POOL_REWARD_ACCOUNT_STEP_INVALID,
+};
+
+static void handlePoolRewardAccount_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = handlePoolRewardAccount_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_POOL_REWARD_ACCOUNT_STEP_DISPLAY_PATH) {
+		ASSERT(subctx->stateData.poolRewardAccount.keyReferenceType == KEY_REFERENCE_PATH);
+
+		ui_displayPathScreen(
+		        "Reward account",
```

Comment:
for the sake of consistency, I'd show the reward account in the same way as for ordinary users, i.e. its ordinal number, followed by the path